ext/json/json_encoder.c

```diff
@@ -531,19 +531,22 @@ zend_result php_json_escape_string(
 static zend_result php_json_encode_serializable_object(smart_str *buf, zval *val, int options, php_json_encoder *encoder) /* {{{ */
 {
 	zend_class_entry *ce = Z_OBJCE_P(val);
-	HashTable* myht = Z_OBJPROP_P(val);
```

Comment:
not sure if I understand the question but the point is to add the guard property for any object that implements JsonSerializable or have __debugInfo so we can distinguish between protections (that's necessary so we can use var_dump in jsonSerialize).

Reply:
I understand, though you mentioned that the main motivation was avoiding building of the properties table. It seemed to me that this may also have been achieved by setting the `GC_PROTECTED` flag on the object instead. Anyway, it doesn't really matter as this solution solves more problems than that.

Reply:
Because that object protection flag is also used by var_dump so the following and other included tests won't work. The easiest one to see it is probably: https://github.com/php/php-src/blob/a037c5f3f2a2e5edde73ad00700f28d75593a451/ext/json/tests/json_encode_recursion_02.phpt . If you use just `GC_PROTECTED`, then recursion check for var_dump will fail because it is already protected by `jsonSerialize`. So those call need to be differentiated and that's why we need more flags. We can't add more to object so Dmitry suggested this approach instead which seems to work.

Reply:
I didn't know we're using the CG flags of objects vs property table for different purposes already. Thanks for clarifying.